modules/gdscript/gdscript_compiler.cpp

```diff
@@ -1200,13 +1262,21 @@ GDScriptCodeGenerator::Address GDScriptCompiler::_parse_expression(CodeGen &code
 					to_assign = assigned_value;
 				}
 
-				GDScriptDataType assign_type = _gdtype_from_datatype(assignment->assignee->get_datatype(), codegen.script);
-
 				if (has_setter && !is_in_setter) {
 					// Call setter.
 					Vector<GDScriptCodeGenerator::Address> args;
 					args.push_back(to_assign);
-					gen->write_call(GDScriptCodeGenerator::Address(), GDScriptCodeGenerator::Address(GDScriptCodeGenerator::Address::SELF), setter_function, args);
+					GDScriptCodeGenerator::Address call_base = is_static ? GDScriptCodeGenerator::Address(GDScriptCodeGenerator::Address::CLASS) : GDScriptCodeGenerator::Address(GDScriptCodeGenerator::Address::SELF);
+					gen->write_call(GDScriptCodeGenerator::Address(), call_base, setter_function, args);
+				} else if (is_static) {
+					GDScriptCodeGenerator::Address temp = codegen.add_temporary(static_var_data_type);
+					if (assignment->use_conversion_assign) {
+						gen->write_assign_with_conversion(temp, to_assign);
+					} else {
+						gen->write_assign(temp, to_assign);
+					}
+					gen->write_set_static_variable(temp, static_var_class, static_var_index);
+					gen->pop_temporary();
```

Comment:
This makes me worry about performance (the need for intermediate temporaries) and maintainability (with this approach, we have to duplicate a lot of tests for static variables that we already have for non-static variables). In this sense, addressing modes is better than separate opcodes.

Since we can't get a contiguous array, maybe we could have a dynamically sized array (or a two-dimensional array if we assume that in the future we might add something else after static arrays)? The address already consists of two parts: the index of the nested array in `variant_addresses` and the index of the element in the nested array.

The compiler could take into account the depth of class inheritance, instead of storing pointers to scripts. A script that doesn't have a base script has a depth of 0, then 1, 2, 3, and so on.

Reply:
Done: https://github.com/dalexeev/godot/commit/14efadf3888f1011fc384280a1de90099f3f75b8

Reply:
I'm not sure about that. Sure it's slower having a two-step approach, but generating this address array dynamically every time seems to be worse for me. Regarding tests, we will need tests for static variables a anyway, so I don't think it makes much of a difference (IMO testing for generating a particular addressing mode is as important as testing for generating opcodes).